README.rst
================================================
minFraud Score, Insights, and Factors Python API
================================================

Description
-----------

This package provides an API for the `MaxMind minFraud Score, Insights, Factors
web services <https://dev.maxmind.com/minfraud/>`_.

Installation
------------

To install the ``minfraud`` module, type:

.. code-block:: bash

    $ pip install minfraud

If you are not able to use pip, you may also use easy_install from the
source directory:

.. code-block:: bash

    $ easy_install .

Documentation
-------------

Complete API documentation is available on `Read the Docs
<http://minfraud.readthedocs.io/>`_.

Usage
-----

To use this API, create a new ``minfraud.Client`` object. The constructor
takes your MaxMind user ID and license key:

.. code-block:: pycon

    >>> client = Client(42, 'licensekey')

The Factors service is called with the ``factors()`` method:

.. code-block:: pycon

    >>> factors = client.factors({'device': {'ip_address': '81.2.69.160'}})

The Insights service is called with the ``insights()`` method:

.. code-block:: pycon

    >>> insights = client.insights({'device': {'ip_address': '81.2.69.160'}})

The Score web service is called with the ``score()`` method:

.. code-block:: pycon

    >>> insights = client.insights({'device': {'ip_address': '81.2.69.160'}})

Each of these methods takes a dictionary representing the transaction to be sent
to the web service. The structure of this dictionary should be in `the format
specified in the REST API documentation
<https://dev.maxmind.com/minfraud/#Request_Body>`_.
The ``ip_address`` in the ``device`` sub-dictionary is required. All other
fields are optional.

Assuming validation has not been disabled, before sending the transaction to
the web service, the transaction dictionary structure and content will be
validated. If validation fails, a ``minfraud.InvalidRequestError``
will be raised.

If the dictionary is valid, a request will be made to the web service. If the
request succeeds, a model object for the service response will be returned.
If the request fails, one of the errors listed below will be raised.

Errors
------

The possible errors are:

* ``minfraud.AuthenticationError`` - This will be raised when the server
  is unable to authenticate the request, e.g., if the license key or user ID
  is invalid.
* ``minfraud.InsufficientFundsError`` - This will be raised when `your
  account <https://www.maxmind.com/en/account>`_ is out of funds.
* ``minfraud.InvalidRequestError`` - This will be raised when the server
  rejects the request as invalid for another reason, such as a missing or
  reserved IP address. It is also raised if validation of the request before
  it is sent to the server fails.
* ``minfraud.HttpError`` - This will be raised when an unexpected HTTP
  error occurs such as a firewall interfering with the request to the server.
* ``minfraud.MinFraudError`` - This will be raised when some other error
  occurs such as unexpected content from the server. This also serves as the
  base class for the above errors.

Example
-------

.. code-block:: pycon

    >>> from minfraud import Client
    >>>
    >>> client = Client(42, 'licensekey')
    >>>
    >>> request = {
    >>>     'device': {
    >>>         'ip_address': '81.2.69.160',
    >>>         'accept_language': 'en-US,en;q=0.8',
    >>>         'user_agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.89 Safari/537.36'
    >>>     },
    >>>     'event': {
    >>>         'shop_id': 's2123',
    >>>         'type': 'purchase',
    >>>         'transaction_id': 'txn3134133',
    >>>         'time': '2014-04-12T23:20:50.052+00:00'
    >>>     },
    >>>     'account': {
    >>>         'user_id': '3132',
    >>>         'username_md5': '570a90bfbf8c7eab5dc5d4e26832d5b1'
    >>>     },
    >>>     'email': {
    >>>         'address': '977577b140bfb7c516e4746204fbdb01',
    >>>         'domain': 'maxmind.com'
    >>>     },
    >>>     'billing': {
    >>>         'first_name': 'Jane'
    >>>         'last_name': 'Doe',
    >>>         'company': 'Company',
    >>>         'address': '101 Address Rd.',
    >>>         'address_2': 'Unit 5',
    >>>         'city': 'Hamden',
    >>>         'region': 'CT',
    >>>         'country': 'US',
    >>>         'postal': '06510',
    >>>         'phone_country_code': '1',
    >>>         'phone_number': '323-123-4321',
    >>>     },
    >>>     'shipping': {
    >>>         'first_name': 'John'
    >>>         'last_name': 'Doe',
    >>>         'company': 'ShipCo',
    >>>         'address': '322 Ship Addr. Ln.',
    >>>         'address_2': 'St. 43',
    >>>         'city': 'New Haven',
    >>>         'region': 'CT',
    >>>         'country': 'US',
    >>>         'postal': '06510',
    >>>         'phone_country_code': '1',
    >>>         'phone_number': '403-321-2323',
    >>>         'delivery_speed': 'same_day',
    >>>     },
    >>>     'credit_card': {
    >>>         'bank_phone_country_code': '1',
    >>>         'avs_result': 'Y',
    >>>         'bank_phone_number': '800-342-1232',
    >>>         'last_4_digits': '7643',
    >>>         'cvv_result': 'N',
    >>>         'bank_name': 'Bank of No Hope',
    >>>         'issuer_id_number': '323132'
    >>>     },
    >>>     'payment': {
    >>>         'decline_code': 'invalid number',
    >>>         'was_authorized': False,
    >>>         'processor': 'stripe'
    >>>     },
    >>>     'shopping_cart': [{
    >>>         'category': 'pets',
    >>>         'quantity': 2,
    >>>         'price': 20.43,
    >>>         'item_id': 'lsh12'
    >>>     }, {
    >>>         'category': 'beauty',
    >>>         'quantity': 1,
    >>>         'price': 100.0,
    >>>         'item_id': 'ms12'
    >>>     }],
    >>>     'order': {
    >>>         'affiliate_id': 'af12',
    >>>         'referrer_uri': 'http://www.amazon.com/',
    >>>         'subaffiliate_id': 'saf42',
    >>>         'discount_code': 'FIRST',
    >>>         'currency': 'USD',
    >>>         'amount': 323.21
    >>>      },
    >>>     'custom_inputs': {
    >>>         'section': 'news',
    >>>         'number_of_previous_purchases': 19,
    >>>         'discount': 3.2,
    >>>         'previous_user': True
    >>>     }
    >>> }
    >>>
    >>> client.score(request)
    Score(...)
    >>>
    >>> client.insights(request)
    Insights(...)
    >>>
    >>> client.factors(request)
    Factors(...)

Requirements
------------

This code requires Python 2.6+ or 3.3+. Older versions are not supported.
This library has been tested with CPython and PyPy.

Versioning
----------

The minFraud Python API uses `Semantic Versioning <http://semver.org/>`_.

Support
-------

Please report all issues with this code using the `GitHub issue tracker
<https://github.com/maxmind/minfraud-api-python/issues>`_.

If you are having an issue with a MaxMind service that is not specific to the
client API, please contact `MaxMind support <http://www.maxmind.com/en/support>`_
for assistance.

Copyright and License
---------------------

This software is Copyright © 2015-2016 by MaxMind, Inc.

This is free software, licensed under the Apache License, Version 2.0.


HISTORY.rst
.. :changelog:

History
-------

1.4.0 (2017-XX-XX)
++++++++++++++++++

* Added support for custom inputs. You may set up custom inputs from your
  account portal.

1.3.2 (2016-12-08)
++++++++++++++++++

* Recent releases of ``requests`` (2.12.2 and 2.12.3) require that the
  username for basic authentication be a string or bytes. The documentation
  for this module uses an integer for the ``user_id``, which will break with
  these ``requests`` versions. The ``user_id`` is now converted to bytes
  before being passed to ``requests``.
* Fixed test breakage on 3.6.

1.3.1 (2016-11-22)
++++++++++++++++++

* Fixed ``setup.py`` on Python 2.

1.3.0 (2016-11-22)
++++++++++++++++++

* The disposition was added to the minFraud response models. This is used to
  return the disposition of the transaction as set by the custom rules for the
  account.
* Fixed package's long description for display on PyPI.

1.2.0 (2016-11-14)
++++++++++++++++++

* Allow ``/credit_card/token`` input.

1.1.0 (2016-10-10)
++++++++++++++++++

* Added the following new values to the ``/event/type`` validation:
  ``email_change`` and ``password_reset``.

1.0.0 (2016-09-15)
++++++++++++++++++

* Added the following new values to the ``/payment/processor`` validation:
  ``concept_payments``, ``ecomm365``, ``orangepay``, and ``pacnet_services``.
* `ipaddress` is now used for IP validation on Python 2 instead of `ipaddr`.

0.5.0 (2016-06-08)
++++++++++++++++++

* BREAKING CHANGE: ``credits_remaining`` has been removed from the web service
  response model and has been replaced by ``queries_remaining``.
* Added ``queries_remaining`` and ``funds_remaining``. Note that
  ``funds_remaining`` will not be returned by the web service until our new
  credit system is in place.
* ``confidence`` and ``last_seen`` were added to the ``Device`` response
  model.

0.4.0 (2016-05-23)
++++++++++++++++++

* Added support for the minFraud Factors.
* Added IP address risk to the minFraud Score model.
* Added the following new values to the ``/payment/processor`` validation:
  ``ccnow``, ``dalpay``, ``epay`` (replaces ``epayeu``), ``payplus``,
  ``pinpayments``, ``quickpay``, and ``verepay``.
* A ``PERMISSION_REQUIRED`` error will now throw a ``PermissionRequiredError``
  exception.

0.3.0 (2016-01-20)
++++++++++++++++++

* Added support for new minFraud Insights outputs. These are:
     * ``/credit_card/brand``
     * ``/credit_card/type``
     * ``/device/id``
     * ``/email/is_free``
     * ``/email/is_high_risk``
* ``input`` on the ``Warning`` response model has been replaced with
  ``input_pointer``. The latter is a JSON pointer to the input that
  caused the warning.

0.2.0 (2015-08-10)
++++++++++++++++++

* Added ``is_gift`` and ``has_gift_message`` to `order` input dictionary
  validation.
* Request keys with ``None`` values are no longer validated or sent to the
  web service.

0.1.0 (2015-06-29)
++++++++++++++++++

* First beta release.

0.0.1 (2015-06-19)
++++++++++++++++++

* Initial release.


tests/test_validation.py
from decimal import Decimal
import sys
from voluptuous import MultipleInvalid

from minfraud.validation import validate_transaction

if sys.version_info[:2] == (2, 6):
    import unittest2 as unittest
else:
    import unittest

if sys.version_info[0] == 2:
    unittest.TestCase.assertRaisesRegex = unittest.TestCase.assertRaisesRegexp
    unittest.TestCase.assertRegex = unittest.TestCase.assertRegexpMatches


class ValidationBase(object):
    def setup_transaction(self, transaction):
        if 'device' not in transaction:
            transaction['device'] = {}

        if 'ip_address' not in transaction['device']:
            transaction['device']['ip_address'] = '1.1.1.1'

    def check_invalid_transaction(self, transaction):
        self.setup_transaction(transaction)
        with self.assertRaises(
                MultipleInvalid, msg='{0!s} is invalid'.format(transaction)):
            validate_transaction(transaction)

    def check_transaction(self, transaction):
        self.setup_transaction(transaction)
        try:
            validate_transaction(transaction)
        except MultipleInvalid as e:
            self.fail('MultipleInvalid {0} thrown for {1}'.format(e.msg,
                                                                  transaction))

    def check_str_type(self, object, key):
        self.check_transaction({object: {key: 'string'}})
        self.check_invalid_transaction({object: {key: 12}})

    def check_positive_number(self, f):
        for good in (1, 1.1, Decimal('1.1')):
            self.check_transaction(f(good))
        for bad in ('1.2', '1', -1, -1.1, 0):
            self.check_invalid_transaction(f(bad))

    def check_bool(self, object, key):
        for good in (True, False):
            self.check_transaction({object: {key: good}})
        for bad in ('', 0, 'True'):
            self.check_invalid_transaction({object: {key: bad}})


class TestAccount(unittest.TestCase, ValidationBase):
    def test_account_user_id(self):
        self.check_transaction({'account': {'user_id': 'usr'}})

    def test_account_username_md5(self):
        self.check_transaction({
            'account': {
                'username_md5': '14c4b06b824ec593239362517f538b29'
            }
        })

    def test_invalid_account_username_md5s(self):
        self.check_invalid_transaction({
            'account': {
                'username_md5': '14c4b06b824ec593239362517f538b2'
            }
        })
        self.check_invalid_transaction({
            'account': {
                'username_md5': '14c4b06b824ec593239362517f538b29a'
            }
        })


class AddressBase(ValidationBase):
    def test_strings(self):
        for key in ('first_name', 'last_name', 'company', 'address',
                    'address_2', 'city', 'postal', 'phone_number'):
            self.check_str_type(self.type, key)

    def test_region(self):
        for region in ('A', 'AA', 'AAA', 'ZZZZ'):
            self.check_transaction({self.type: {'region': region}})
        for invalid in ('', 'AAAAA', 1, 'aaa'):
            self.check_invalid_transaction({self.type: {'region': invalid}})

    def test_country(self):
        for country in ('US', 'CA', 'GB'):
            self.check_transaction({self.type: {'country': country}})
        for invalid in ('', 'U1', 'USA', 1, '11', 'us'):
            self.check_invalid_transaction({self.type: {'country': invalid}})

    def test_phone_country_code(self):
        for code in (1, '1', '2341'):
            self.check_transaction({self.type: {'phone_country_code': code}})
        for invalid in ('', '12345', 'U'):
            self.check_invalid_transaction({
                self.type: {
                    'phone_country_code': invalid
                }
            })


class TestBillingAddress(unittest.TestCase, AddressBase):
    type = 'billing'


class TestShippingAddress(unittest.TestCase, AddressBase):
    type = 'shipping'

    def test_delivery_speed(self):
        for speed in ('same_day', 'overnight', 'expedited', 'standard'):
            self.check_transaction({self.type: {'delivery_speed': speed}})
        for invalid in ('fast', 'slow', ''):
            self.check_invalid_transaction({
                self.type: {
                    'delivery_speed': invalid
                }
            })


class TestCreditCard(ValidationBase, unittest.TestCase):
    def test_issuer_id_number(self):
        for iin in ('123456', '532313'):
            self.check_transaction({'credit_card': {'issuer_id_number': iin}})
        for invalid in ('12345', '1234567', 123456, '12345a'):
            self.check_invalid_transaction({
                'credit_card': {
                    'issuer_id_number': invalid
                }
            })

    def test_last_4_digits(self):
        for iin in ('1234', '9323'):
            self.check_transaction({'credit_card': {'last_4_digits': iin}})
        for invalid in ('12345', '123', 1234, '123a'):
            self.check_invalid_transaction({
                'credit_card': {
                    'last_4_digits': invalid
                }
            })

    def test_bank_name(self):
        self.check_str_type('credit_card', 'bank_name')

    def test_bank_phone_number(self):
        self.check_str_type('credit_card', 'bank_phone_number')

    def test_phone_country_code(self):
        for code in (1, '1', '2341'):
            self.check_transaction({
                'credit_card': {
                    'bank_phone_country_code': code
                }
            })
        for invalid in ('', '12345', 'U'):
            self.check_invalid_transaction({
                'credit_card': {
                    'bank_phone_country_code': invalid
                }
            })

    def test_avs_and_cvv(self):
        for key in ('avs_result', 'cvv_result'):
            for code in ('1', 'A'):
                self.check_transaction({'credit_card': {key: code}})
            for invalid in ('', '12'):
                self.check_invalid_transaction({
                    'credit_card': {
                        'credit_card': invalid
                    }
                })

    def test_token(self):
        for token in ('123456abc1245', '\x21', '1' * 20):
            self.check_transaction({'credit_card': {'token': token}})
        for invalid in ('\x20', '123456', 'x' * 256):
            self.check_invalid_transaction({'credit_card': {'token': invalid}})


class TestCustomInputs(ValidationBase, unittest.TestCase):
    def test_valid_inputs(self):
        self.check_transaction({
            'custom_inputs': {
                'string_input_1': 'test string',
                'int_input': 19,
                'float_input': 3.2,
                'bool_input': True
            }
        })

    def test_invalid(self):
        for invalid in (
            {'InvalidKey': 1},
            {'too_long': 'x' * 256},
            {'has_newline': 'test\n'},
            {'too_big': 1<<53},
            {'too_small': -(1<<53)},
            {'too_big_float': float(1<<53)}
            ): # yapf: disable
            self.check_invalid_transaction({'custom_inputs': invalid})


class TestDevice(ValidationBase, unittest.TestCase):
    def test_ip_address(self):
        for ip in ('1.2.3.4', '2001:db8:0:0:1:0:0:1', '::FFFF:1.2.3.4'):
            self.check_transaction({'device': {'ip_address': ip}})
        for invalid in ('1.2.3.', '299.1.1.1', '::AF123'):
            self.check_invalid_transaction({'device': {'ip_address': invalid}})

    def test_missing_ip(self):
        with self.assertRaises(MultipleInvalid):
            validate_transaction({'device': {}})

    def test_missing_device(self):
        with self.assertRaises(MultipleInvalid):
            validate_transaction({})

    def test_user_agent(self):
        self.check_str_type('device', 'user_agent')

    def test_accept_language(self):
        self.check_str_type('device', 'accept_language')


class TestEmail(ValidationBase, unittest.TestCase):
    def test_address(self):
        for good in ('test@maxmind.com', '977577b140bfb7c516e4746204fbdb01'):
            self.check_transaction({'email': {'address': good}})
        for bad in ('not.email', '977577b140bfb7c516e4746204fbdb0',
                    '977577b140bfb7c516e4746204fbdb012'):
            self.check_invalid_transaction({'email': {'address': bad}})

    def test_domain(self):
        for good in ('maxmind.com', 'www.bbc.co.uk'):
            self.check_transaction({'email': {'domain': good}})
        for bad in ('bad ', ' bad.com'):
            self.check_invalid_transaction({'email': {'domain': bad}})


class TestEvent(ValidationBase, unittest.TestCase):
    def test_transaction(self):
        self.check_str_type('event', 'transaction_id')

    def test_shop_id(self):
        self.check_str_type('event', 'shop_id')

    def test_time(self):
        for good in ('2015-05-08T16:07:56+00:00', '2015-05-08T16:07:56Z'):
            self.check_transaction({'event': {'time': good}})
        for bad in ('2015-05-08T16:07:56', '2015-05-08 16:07:56Z'):
            self.check_invalid_transaction({'event': {'time': bad}})

    def test_type(self):
        for good in ('account_creation', 'account_login', 'email_change',
                     'password_reset', 'purchase', 'recurring_purchase',
                     'referral', 'survey'):
            self.check_transaction({'event': {'type': good}})
        for bad in ('bad', 1, ''):
            self.check_invalid_transaction({'event': {'type': bad}})


class TestOrder(ValidationBase, unittest.TestCase):
    def test_amount(self):
        self.check_positive_number(lambda v: {'order': {'amount': v}})

    def test_currency(self):
        for good in ('USD', 'GBP'):
            self.check_transaction({'order': {'currency': good}})
        for bad in ('US', 'US1', 'USDD', 'usd'):
            self.check_invalid_transaction({'order': {'currency': bad}})

    def test_discount_code(self):
        self.check_str_type('order', 'discount_code')

    def test_affiliate_id(self):
        self.check_str_type('order', 'affiliate_id')

    def test_subaffiliate_id(self):
        self.check_str_type('order', 'subaffiliate_id')

    def test_is_gift(self):
        self.check_bool('order', 'is_gift')

    def test_has_gift_message(self):
        self.check_bool('order', 'has_gift_message')

    def test_referrer_uri(self):
        for good in ('http://www.mm.com/fadsf', 'https://x.org/'):
            self.check_transaction({'order': {'referrer_uri': good}})
        for bad in ('ftp://a.com/', 'www.mm.com'):
            self.check_invalid_transaction({'order': {'referrer_uri': bad}})


class TestPayment(ValidationBase, unittest.TestCase):
    def test_processor(self):
        for good in ('adyen', 'stripe'):
            self.check_transaction({'payment': {'processor': good}})
        for bad in ('notvalid', ' stripe'):
            self.check_invalid_transaction({'payment': {'processor': bad}})

    def test_was_authorized(self):
        self.check_bool('payment', 'was_authorized')

    def test_decline_code(self):
        self.check_str_type('payment', 'decline_code')


class TestShoppingCart(ValidationBase, unittest.TestCase):
    def test_category(self):
        self.check_transaction({'shopping_cart': [{'category': 'cat'}]})

    def test_item_id(self):
        self.check_transaction({'shopping_cart': [{'item_id': 'cat'}]})

    def test_amount(self):
        self.check_positive_number(lambda v: {'shopping_cart': [{'price': v}]})

    def test_quantity(self):
        for good in (1, 1000):
            self.check_transaction({'shopping_cart': [{'quantity': good}]})
        for bad in (1.1, -1, 0):
            self.check_invalid_transaction({
                'shopping_cart': [{
                    'quantity': bad
                }]
            })


minfraud/validation.py
"""This is an internal module used for validating the minFraud request."""

import re
import sys
from decimal import Decimal

from geoip2.compat import compat_ip_address
from strict_rfc3339 import validate_rfc3339
from validate_email import validate_email
from voluptuous import All, Any, In, Match, Range, Required, Schema
import rfc3987  # pylint:disable=import-error
"""
Internal code for validating the transaction dictionary.

This code is only intended for internal use and is subject to change in ways
that may break any direct use of it.

"""

# Pylint doesn't like the private function type naming for the callable
# objects below. Given the consistent use of them, the current names seem
# preferable to blindly following pylint.
#
# pylint: disable=invalid-name,redefined-builtin,redefined-variable-type,undefined-variable

if sys.version_info[0] >= 3:
    _unicode_or_printable_ascii = str
    long = int
else:
    _unicode_or_printable_ascii = Any(unicode, Match(r'^[\x20-\x7E]*$'))
# pylint: enable=redefined-variable-type

_any_string = Any(_unicode_or_printable_ascii, str)

_custom_input_key = All(_any_string, Match(r'^[a-z0-9_]{1,25}$'))

_custom_input_value = Any(All(_any_string, Match(r'^[^\n]{1,255}\Z')),
                          All(Any(int, float, long),
                              Range(
                                  min=-(1 << 53),
                                  max=1 << 53,
                                  min_included=False,
                                  max_included=False)),
                          bool)

_md5 = All(_any_string, Match(r'^[0-9A-Fa-f]{32}$'))

_country_code = All(_any_string, Match(r'^[A-Z]{2}$'))

_telephone_country_code = Any(All(_any_string, Match(r'^[0-9]{1,4}$')),
                              All(int, Range(
                                  min=1, max=9999)))

_subdivision_iso_code = All(_any_string, Match(r'^[0-9A-Z]{1,4}$'))


def _ip_address(s):
    # ipaddress accepts numeric IPs, which we don't want. Pylint on Python 3
    # doesn't like "unicode"
    #
    # pylint: disable=undefined-variable
    if (isinstance(s, str) or isinstance(s, unicode)) \
            and not re.match(r'^\d+$', s):
        return str(compat_ip_address(s))
    raise ValueError


def _email_or_md5(s):
    if validate_email(s) or re.match(r'^[0-9A-Fa-f]{32}$', s):
        return s
    raise ValueError


# based off of:
# http://stackoverflow.com/questions/2532053/validate-a-hostname-string
def _hostname(hostname):
    if len(hostname) > 255:
        raise ValueError
    allowed = re.compile(r"(?!-)[A-Z\d-]{1,63}(?<!-)$", re.IGNORECASE)
    if all(allowed.match(x) for x in hostname.split(".")):
        return hostname
    raise ValueError


_delivery_speed = In(['same_day', 'overnight', 'expedited', 'standard'])

_address = {
    'address': _unicode_or_printable_ascii,
    'address_2': _unicode_or_printable_ascii,
    'city': _unicode_or_printable_ascii,
    'company': _unicode_or_printable_ascii,
    'country': _country_code,
    'first_name': _unicode_or_printable_ascii,
    'last_name': _unicode_or_printable_ascii,
    'phone_country_code': _telephone_country_code,
    'phone_number': _unicode_or_printable_ascii,
    'postal': _unicode_or_printable_ascii,
    'region': _subdivision_iso_code,
}

_shipping_address = _address.copy()

_shipping_address['delivery_speed'] = _delivery_speed

_payment_processor = In([
    'adyen',
    'altapay',
    'amazon_payments',
    'authorizenet',
    'balanced',
    'beanstream',
    'bluepay',
    'braintree',
    'ccnow',
    'chase_paymentech',
    'cielo',
    'collector',
    'compropago',
    'concept_payments',
    'conekta',
    'cuentadigital',
    'dalpay',
    'dibs',
    'digital_river',
    'ecomm365',
    'elavon',
    'epay',
    'eprocessing_network',
    'eway',
    'first_data',
    'global_payments',
    'ingenico',
    'internetsecure',
    'intuit_quickbooks_payments',
    'iugu',
    'mastercard_payment_gateway',
    'mercadopago',
    'merchant_esolutions',
    'mirjeh',
    'mollie',
    'moneris_solutions',
    'nmi',
    'openpaymx',
    'optimal_payments',
    'orangepay',
    'other',
    'pacnet_services',
    'payfast',
    'paygate',
    'payone',
    'paypal',
    'payplus',
    'paystation',
    'paytrace',
    'paytrail',
    'payture',
    'payu',
    'payulatam',
    'pinpayments',
    'princeton_payment_solutions',
    'psigate',
    'qiwi',
    'quickpay',
    'raberil',
    'rede',
    'redpagos',
    'rewardspay',
    'sagepay',
    'simplify_commerce',
    'skrill',
    'smartcoin',
    'sps_decidir',
    'stripe',
    'telerecargas',
    'towah',
    'usa_epay',
    'verepay',
    'vindicia',
    'virtual_card_services',
    'vme',
    'worldpay',
])

_single_char = Match('^[A-Za-z0-9]$')

_iin = Match('^[0-9]{6}$')

_credit_card_last_4 = Match('^[0-9]{4}$')


def _credit_card_token(s):
    if re.match('^[\x21-\x7E]{1,255}$', s) and not re.match('^[0-9]{1,19}$',
                                                            s):
        return s
    raise ValueError


def _rfc3339_datetime(s):
    if validate_rfc3339(s):
        return s
    raise ValueError


_event_type = In([
    'account_creation',
    'account_login',
    'email_change',
    'password_reset',
    'purchase',
    'recurring_purchase',
    'referral',
    'survey',
])

_currency_code = Match('^[A-Z]{3}$')

_price = All(Any(float, int, Decimal), Range(min=0, min_included=False))


def _uri(s):
    if rfc3987.parse(s).get('scheme') in ['http', 'https']:
        return s
    raise ValueError


validate_transaction = Schema(
    {
        'account': {
            'user_id': _unicode_or_printable_ascii,
            'username_md5': _md5,
        },
        'billing': _address,
        'payment': {
            'processor': _payment_processor,
            'was_authorized': bool,
            'decline_code': _unicode_or_printable_ascii,
        },
        'credit_card': {
            'avs_result': _single_char,
            'bank_name': _unicode_or_printable_ascii,
            'bank_phone_country_code': _telephone_country_code,
            'bank_phone_number': _unicode_or_printable_ascii,
            'cvv_result': _single_char,
            'issuer_id_number': _iin,
            'last_4_digits': _credit_card_last_4,
            'token': _credit_card_token,
        },
        'custom_inputs': {
            _custom_input_key: _custom_input_value
        },
        Required('device'): {
            'accept_language': _unicode_or_printable_ascii,
            Required('ip_address'): _ip_address,
            'user_agent': _unicode_or_printable_ascii
        },
        'email': {
            'address': _email_or_md5,
            'domain': _hostname,
        },
        'event': {
            'shop_id': _unicode_or_printable_ascii,
            'time': _rfc3339_datetime,
            'type': _event_type,
            'transaction_id': _unicode_or_printable_ascii,
        },
        'order': {
            'affiliate_id': _unicode_or_printable_ascii,
            'amount': _price,
            'currency': _currency_code,
            'discount_code': _unicode_or_printable_ascii,
            'has_gift_message': bool,
            'is_gift': bool,
            'referrer_uri': _uri,
            'subaffiliate_id': _unicode_or_printable_ascii,
        },
        'shipping': _shipping_address,
        'shopping_cart': [{
            'category': _unicode_or_printable_ascii,
            'item_id': _unicode_or_printable_ascii,
            'price': _price,
            'quantity': All(int, Range(min=1)),
        }, ],
    }, )
